### PR TITLE
Switch from Peewee to MongoDB

### PR DESCRIPTION
**Pull request by davidbgk**
💣

### CODE REVIEW

File: templates/profile.html [hunk 1]
```diff
@@ -80,7 +80,7 @@
         'website': fields.find(f => f.name === 'website').value.trim(),
         'about': fields.find(f => f.name === 'about').value.trim(),
         'socials': {},
-        'is_editor': document.querySelector('[name=editor]').checked,
+        // 'is_editor': document.querySelector('[name=editor]').checked,
```
Comment (vinyll) [inline templates/profile.html:83]: We need to reactive this right after

File: cafebabel/api/views.py [hunk 2]
```diff
@@ -1,36 +1,35 @@
 from flask import request, jsonify
 from flask_login import current_user, login_required
+from mongoengine.queryset.visitor import Q
```
Comment (vinyll) [inline cafebabel/api/views.py]: remove me

File: cafebabel/api/views.py [hunk 3]
```diff
@@ -1,36 +1,35 @@
...
     data = request.get_json()
     profile_data = {k: data[k]
                     for k in ['name', 'socials', 'website', 'about']}
-    (UserProfile.update(**profile_data)
-                .where(UserProfile.user_id == current_user.id).execute())
-
-    user = User.get(id=current_user.id)
-    editor = Role.get(name='editor')
+    UserProfile.objects(user=current_user.id).modify(**profile_data)
+    editor = Role.objects(name='editor').get()
```
Comment (vinyll) [inline cafebabel/api/views.py]: `editor = Role.objects.get(name='editor')`

File: cafebabel/core/commands.py [hunk 4]
```diff
@@ -1,31 +1,24 @@
...
+from .. import app
+from ..users.models import Role, User, UserProfile, user_datastore
 
 
 def _initdb():
     _dropdb()
-    create_model_tables(tables, fail_silently=True)
-    Role.create(name='editor')
-    admin_role = Role.create(name='admin')
+    Role(name='editor').save()
```
Comment (vinyll) [inline cafebabel/core/commands.py]: Note you can also do `Role.create(name='editor')`

File: cafebabel/tests/test_profile.py [hunk 5]
```diff
@@ -4,7 +4,7 @@
 
 
 def test_confirm_user_creates_profile(app):
-    user = User.create(email='testy@tester.local', password='secret')
+    user = User(email='test_user@example.com', password='secret').save()
     with app.app_context():
         confirm_user(user)
-    assert UserProfile.get(user=user)
+    assert UserProfile.objects(user=user).get()
```
Comment (vinyll) [inline cafebabel/tests/test_profile.py]: `.get(**kwargs)`

File: cafebabel/tests/test_users.py [hunk 6]
```diff
@@ -10,8 +10,8 @@ def test_user_has_no_default_roles(user):
 
 
 def test_user_add_custom_role(user):
-    editor = Role.get(name='editor')
-    UserRoles.create(user=user, role=editor)
+    editor = Role.objects(name='editor').get()
```
Comment (vinyll) [inline cafebabel/tests/test_users.py]: `.get(**kwargs)`

File: cafebabel/tests/test_users.py [hunk 7]
```diff
@@ -26,3 +26,11 @@ def test_admin_has_all_roles(admin):
...
     response = client.get('/profile/')
     assert response.status_code == HTTPStatus.FOUND
+
+
+def test_authenticated_user_can_access_login_required_page(client, user):
+    data = dict(email=user.email, password='secret', remember='y')
+    response = client.post('/login', data=data)
+    assert response.status_code == HTTPStatus.OK
+    response = client.get('/profile/')
+    # assert response.status_code == HTTPStatus.OK
```
Comment (vinyll) [inline cafebabel/tests/test_users.py]: ❓

File: cafebabel/users/models.py [hunk 8]
```diff
@@ -36,31 +41,28 @@ def has_role(self, role, or_admin=True):
 
     @property
     def profile(self):
-        return self.profiles.get()
-
+        return UserProfile.objects(user=self).get()
```
Comment (vinyll) [inline cafebabel/users/models.py]: `.get(**kwargs)`

File: cafebabel/users/models.py [hunk 9]
```diff
@@ -36,31 +41,28 @@ def has_role(self, role, or_admin=True):
...
+signals.post_save.connect(User.post_save, sender=User)
 
 
-class UserProfile(db.Model):
-    name = CharField(null=True)
-    user = ForeignKeyField(User, related_name='profiles', on_delete='CASCADE')
-    socials = PickledField(null=True)
-    website = CharField(null=True)
-    about = CharField(null=True)
+class UserProfile(db.Document):
```
Comment (vinyll) [inline cafebabel/users/models.py]: Do we still need this as a document or as a EmbeddedDocument?

File: cafebabel/users/views.py [hunk 10]
```diff
@@ -12,10 +12,10 @@ def profile():
     return render_template('profile.html', user=user, edit=True)
 
 
-@app.route('/profile/<int:id>/')
+@app.route('/profile/<id>/')
 def profile_user(id):
     try:
-        user = User.get(id=id)
+        user = User.objects(id=id).get()
```
Comment (vinyll) [inline cafebabel/users/views.py]: `.get(**kwargs)`

File: conftest.py [hunk 11]
```diff
@@ -16,20 +16,17 @@ def pytest_runtest_teardown():
...
         confirm_user(user)
     return user
 
 
-@pytest.fixture
+@pytest.fixture(scope='session')
 def admin():
-    _initdb()
-    return User.get(email='admin@example.com')
+    return User.objects(email='admin@example.com').get()
```
Comment (vinyll) [inline conftest.py]: `.get(**kwargs)`

File: cafebabel/core/commands.py (hunk 4 above)
Comment (davidbgk) [reply cafebabel/core/commands.py]: It doesn't look to be the case, source?

File: cafebabel/users/models.py (hunk 9 above)
Comment (davidbgk) [reply cafebabel/users/models.py]: Good question, I don't know Mongo well enough to choose, wdYt?

File: cafebabel/core/commands.py (hunk 4 above)
Comment (vinyll) [reply cafebabel/core/commands.py]: Meant `Role.objects.create(name='editor')`
http://docs.mongoengine.org/apireference.html#mongoengine.queryset.QuerySet.create

Comment (vinyll) [reply cafebabel/core/commands.py]: It makes more sense to me to have an EmbeddeDocument as:

- it is fully dependant of the `User` model and should therefore not be a autonomous `Document`
- it has a schema to respect and should therefore not be a json or object field

**Review by vinyll — APPROVED** [review]
🎉

**Pull request merged**